### PR TITLE
feat: next 12.3 & on-demand ISR

### PR DESCRIPTION
**Pull request by BRKalow**
## 🔗 Relevant links

- [Preview link](https://dev-portal-git-BRANCH_NAME-hashicorp.vercel.app/PATH_TO_VIEW) 🔎
- [Asana task](url) 🎟️

## 🗒️ What

See #1035 for more details.

Moves the beta opt-in redirects from our redirects config to middleware. Adds a new script which generates a file, `src/.generated/opt-in-redirects.js` that contains a map of product slugs to RegEx patterns which will match any redirect paths from a product domain.

diff of changes that aren't part of #1035: https://github.com/hashicorp/dev-portal/pull/1121/files/4201158281ad3444aa341def18a4a3cccce67e78..a41f2f95b64b9f7795cc9186ac51dde1711fecff

### CODE REVIEW

File: src/pages/api/revalidate.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
+import type { NextApiRequest, NextApiResponse } from 'next/types'
+import { StatusCodes } from 'http-status-codes'
+import { validateToken } from 'lib/api-validate-token'
+import { cachedGetProductData } from 'lib/get-product-data'
+
+/**
+ * Accepts a POST request with a product slug, triggers revalidation for all of a product's paths
+ * specified in its latest nav data.
+ */
+async function handler(request: NextApiRequest, response: NextApiResponse) {
```
Comment (kendallstrautman) [inline src/pages/api/revalidate.ts:10]: Would we also have a unique `revalidate` handler for tutorials that only handles a single tutorial or collection path?

File: src/pages/api/revalidate.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
+import type { NextApiRequest, NextApiResponse } from 'next/types'
+import { StatusCodes } from 'http-status-codes'
+import { validateToken } from 'lib/api-validate-token'
+import { cachedGetProductData } from 'lib/get-product-data'
+
+/**
+ * Accepts a POST request with a product slug, triggers revalidation for all of a product's paths
```
Comment (kendallstrautman) [inline src/pages/api/revalidate.ts:7]:
```suggestion
 * Accepts a POST request with a product slug, triggers revalidation for all of a product's docs paths
```

File: src/pages/api/revalidate.ts (hunk 1 above)
Comment (BRKalow) [reply src/pages/api/revalidate.ts:10]: I'll probably update this handler to also handle an arbitrary list of paths to revalidate 👍

File: scripts/generate-opt-in-redirect-checks.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	'src',
+	'.generated',
+	'opt-in-redirect-checks.js'
+)
+
+export default async function main() {
+	const result = ['export default {']
+
+	productSlugs.forEach((productSlug) => {
+		if (productSlug === 'hcp' || productSlug === 'terraform') {
```
Comment (kendallstrautman) [inline scripts/generate-opt-in-redirect-checks.ts:17]: Where are we handling the opt in redirects for these two?

Comment (BRKalow) [reply scripts/generate-opt-in-redirect-checks.ts:17]: In `hashicorp/cloud.hashicorp.com` and `hashicorp/terraform-website` 👍

**Review by kendallstrautman — APPROVED** [review]
Thanks for working through this one! Stoked to start using the on-demand ISR

**Pull request merged**